src/java/main/org/apache/zookeeper/server/NettyServerCnxnFactory.java

```diff
@@ -162,7 +162,7 @@ public void messageReceived(ChannelHandlerContext ctx, MessageEvent e)
                             + " from " + ctx.getChannel());
                 }
                 NettyServerCnxn cnxn = (NettyServerCnxn)ctx.getAttachment();
-                synchronized(cnxn) {
+                synchronized(cnxn.getRpcLock()) {
                     processMessage(e, cnxn);
```

Comment:
>> Now one thread can be in processMessage while another thread is getting stats about the connection. Is that ok?

> I believe this is OK. We can receive two concurrent stat commands, but we only process one of them at a time. I'm also not a Netty wizard, so I could be wildly wrong :)

This should be OK since all fields printed by `synchronized void dumpConnectionInfo(PrintWriter pwriter, boolean brief)` are only updated in `synchronized` methods.

Another option would be to put a shared lock on the command's execution (shared with any other 4LW that acquires a lock on a connection other than its own, I THINK `cons` is the only other one)? I think this may impact 4LW performance slightly but it would prevent the need to add another lock to `NettyServerCnxn`. 

I think I like your way better.